src/libp2p_transport_tcp.erl

```diff
@@ -67,16 +68,34 @@ new_connection(Socket) ->
     libp2p_connection:new(?MODULE, #tcp_state{socket=Socket, transport=ranch_tcp}).
 
 -spec dial(string(), [libp2p_swarm:connect_opt()], pos_integer()) -> {ok, libp2p_connection:connection()} | {error, term()}.
-dial(MAddr, _DialOptions, Timeout) ->
+dial(MAddr, DialOptions, Timeout) ->
     case tcp_addr(MAddr) of
         {IP, Port, Type} ->
-            case ranch_tcp:connect(IP, Port, [Type], Timeout) of
+            UniquePort = proplists:get_value(unique_port, DialOptions, false),
```

Comment:
Why not just use `unique` here instead of introducing `unique_port`?

Reply:
Unique makes sure we don't reuse an existing session, unique_port makes sure we don't use reuseport. They are different operations that will have different behaviour depending on the type of NAT. Specifically restricted NAT vs port restricted NAT.

Reply:
ah, so you can have a unique session and yet reuse a port?

Reply:
Yes, which would result in the same IP/port combo on both sides, but a *new* socket being opened. This is sort of the magic of reuseport.